rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+  syntax/semantics to be used. This allows the interpretation of
+  flakes to change in the future. It also enables some evolution of
+  the Nix language; for example, the Nix files in the flake could be
+  parsed using a syntax determined by the edition. The only currently
+  allowed value is `201906`. Nix rejects flakes with an unsupported
+  edition.
+
+* `name`: An identifier for the flake, used to reference it from
+  `inputs`. (E.g. the `nixpkgs` in `inputs.nixpkgs` refers to the
+  identifier of the Nixpkgs flake.)
+
+* `description`: A short description of the flake.
+
+* `inputs`: The dependencies of the flake, as a list of flake
+  references (described below).
+
+* `outputs`: A function that, given an attribute set containing the
+  outputs of each of the input flakes keyed by their identifier,
+  yields the Nix values provided by this flake. Thus, in the example
+  above, `inputs.nixpkgs` contains the result of the call to the
+  `outputs` function of the `nixpkgs` flake, and
+  `inputs.nixpkgs.packages.fuse` refers to the `packages.fuse` output
+  attribute of `nixpkgs`.
+
+  In addition to the outputs of each input, each input in `inputs`
+  also contains some metadata about the inputs. These are:
+
+  * `outPath`: The path in the Nix store of the flake's source
+    tree. This means that you could import Nixpkgs in a more
+    legacy-ish way by writing
+
+        with import inputs.nixpkgs { system = "x86_64-linux"; };
+
+    since `nixpkgs` still contains a `/default.nix`. In this case we
+    bypass its outputs entirely and only use the flake mechanism to
+    get its source tree.
+
+  * `rev`: The commit hash of the flake's Git repository.
+
+  * `revCount`: The number of ancestors of the revision `rev`. This is
+    not available for `github:...` repositories (see below), since
+    they're fetched as tarballs rather than as Git repositories.
+
+  * `lastModified`: The commit time of the revision `rev`, in the
+    format `%Y%m%d%H%M%S` (e.g. `20181231100934`). Unlike `revCount`,
+    this is available for both Git and GitHub repositories, so it's
+    useful for generating (hopefully) monotonically increasing version
+    strings.
+
+  The special input named `self` refers to the outputs and source tree
+  of *this* flake.
+
+  A number of outputs have a special meaning, as discussed below.
+
+## Well-known outputs
+
+A number of outputs have a specific meaning to Nix or other tools like
+Hydra. Currently, these are:
+
+* `packages`: A set of derivations used as a default by most `nix`
+  commands. For example, `nix run nixpkgs:hello` uses the
+  `packages.hello` attribute of the `nixpkgs` flake. It cannot contain
+  any non-derivation attributes. This also means it cannot be a nested
+  set! (The rationale is that supporting nested sets requires Nix to
+  evaluate each attribute in the set, just to discover which packages
+  are provided.)
+
+* `defaultPackage`: A derivation used as a default by most `nix`
```

Comment:
I think the `defaultPackage`/`defaultApp` can lead to confusion. It can be difficult for newcomers to find out if they have to do
```
nix build dwarffs
nix build dwarffs:dwarffs
nix build nixpkgs:dwarffs
```

I think it would be much clearer if we skip the default packages and just always build `dwarffs:dwarffs` or maybe `dwarffs:pkg`. I think single-usage flakes should be the exception rather than the rule anyway. There is a lot of value in the monorepo approach.

Reply:
Weak +1

Reply:
I think it's the other way around: most flakes will be single-package (just like most crates are single-app/library). Flakes like Nixpkgs would be the exception

Reply:
I think it would be bad for UX not to offer the possibility of defaults. Without them, for example, you wouldn't be able to do
```
# nix build
```
anymore. Instead you would have to write
```
# nix build .:<some attribute name that you have to figure out / remember>
```

Reply:
This seems like a feature that we could add later. That way we can see how people actually use it, and whether lots of people want it.

Reply:
I think it’s inevitable that a convention will emerge. If we don’t add it from the start, though, we risk coming up with competing conventions, and it would be nice to avoid that.

Reply:
I already know I want it :)

Reply:
I know I use it :-)